FirebaseStorage/Sources/StorageDownloadTask.swift

```diff
@@ -26,37 +31,173 @@ import FirebaseStorageInternal
+        fetcher = GTMSessionFetcher(downloadResumeData: resumeData)
+        fetcher.comment = "Resuming DownloadTask"
+      } else {
+        fetcher = strongSelf.fetcherService.fetcher(with: request)
+        fetcher.comment = "Resuming DownloadTask"
+      }
+      fetcher.maxRetryInterval = strongSelf.reference.storage.maxDownloadRetryInterval
+
+      if let fileURL = strongSelf.fileURL {
+        // Handle file downloads
+        fetcher.destinationFileURL = fileURL
+        fetcher.downloadProgressBlock = { (bytesWritten: Int64,
+                                           totalBytesWritten: Int64,
+                                           totalBytesExpectedToWrite: Int64) in
+            weakSelf?.state = .progress
+            weakSelf?.progress.completedUnitCount = totalBytesWritten
+            weakSelf?.progress.totalUnitCount = totalBytesExpectedToWrite
+            if let snapshot = weakSelf?.snapshot {
+              weakSelf?.fire(for: .progress, snapshot: snapshot)
+            }
+            weakSelf?.state = .running
+        }
+      } else {
+        // Handle data downloads
+        fetcher.receivedProgressBlock = { (bytesWritten: Int64, totalBytesWritten: Int64) in
+          weakSelf?.state = .progress
+          weakSelf?.progress.completedUnitCount = totalBytesWritten
+          if let totalLength = weakSelf?.fetcher?.response?.expectedContentLength {
+            weakSelf?.progress.totalUnitCount = totalLength
+          }
+          if let snapshot = weakSelf?.snapshot {
+            weakSelf?.fire(for: .progress, snapshot: snapshot)
+          }
+          weakSelf?.state = .running
+        }
+      }
+      strongSelf.fetcher = fetcher
+      strongSelf.fetcherCompletion = { (data: Data?, error: NSError?) in
+        self.fire(for: .progress, snapshot: self.snapshot)
+
+        // Handle potential issues with download
+        if let error = error {
+          self.state = .failed
+          self.error = StorageErrorCode.error(withServerError: error, ref: self.reference)
+          self.fire(for: .failure, snapshot: self.snapshot)
+          self.removeAllObservers()
+          self.fetcherCompletion = nil
+          return
+        }
+        // Download completed successfully, fire completion callbacks
+        self.state = .success
+        if let data = data {
+          self.downloadData = data
+        }
+        self.fire(for: .success, snapshot: self.snapshot)
+        self.removeAllObservers()
+        self.fetcherCompletion = nil
+      }
+      strongSelf.state = .running
+      strongSelf.fetcher?.beginFetch { data, error in
+        let strongSelf = weakSelf
+        if let fetcherCompletion = strongSelf?.fetcherCompletion {
+          fetcherCompletion(data, error as? NSError)
+        }
+      }
+    }
+  }
+
+  internal func cancel(withError error: NSError) {
+    weak var weakSelf = self
+    DispatchQueue.global(qos: .background).async {
+      weakSelf?.state = .cancelled
```

Comment:
Good question. This is a straight port. I suppose that since everything is a message send to weakSelf, if it goes nil, then the rest of the block will be a no-op.